### ISSUE TITLE
Detect infinite recursion (at compile-time) using stable names

### ISSUE THREAD

**Issue by L-as** [Plutonomicon/plutarch-plutus#19]
This will likely heavily slow down the compiler, but it's probably worth it.

**Comment by TotallyNotChase** [comment]
An example that (probably) belongs here-
```
recurH :: Term s PUnit -> Term s PUnit
recurH _ = pif (pcon PTrue) (pcon PUnit) $ recurH $ pcon PUnit

-- >>> compile $ recurH £ pcon PUnit
```

**Comment by L-as** [comment]
A much simpler example is just
```
bad :: Term s a
bad = bad
```

**Comment by AriFordsham** [comment]
How do you do this without solving the halting problem?

**Comment by L-as** [comment]
You can't. You can however catch many of the simple cases, just like Nix does.
If you do `let x = x; in x` in Nix, it's going to give you an error message.

**Comment by L-as** [comment]
@TotallyNotChase Looking back on this almost three months later, has this ever been a problem for you in practice?

Also relevant for @srid and @Geometer1729

**Comment by Geometer1729** [comment]
It hasn't been a problem for me yet.

**Comment by TotallyNotChase** [comment]
Nope. It shouldn't be too common of a mistake to make in general. But I can definitely see the mistake *potentially* being made.

**Comment by L-as** [comment]
That doesn't tell you where the problem is
